packages/plugin-hardhat/src/utils/simulate-deploy.ts

```diff
@@ -0,0 +1,72 @@
+import { fetchOrDeploy, fetchOrDeployAdmin, hashBytecode, logWarning } from '@openzeppelin/upgrades-core';
+import type { ContractFactory } from 'ethers';
+import { FormatTypes } from 'ethers/lib/utils';
+import type { HardhatRuntimeEnvironment } from 'hardhat/types';
+import { getDeployData } from './deploy-impl';
+import { Options } from './options';
+
+// To import an already deployed contract we want to reuse fetchOrDeploy for its ability to validate
+// a deployment and record it in the network file. We are able to do this by "simulating" a deployment:
+// for the "deploy" part we pass a function that simply returns the contract to be imported, rather than
+// actually deploying something.
+
+export async function simulateDeployAdmin(
+  hre: HardhatRuntimeEnvironment,
+  ProxyAdminFactory: ContractFactory,
+  opts: Options,
+  adminAddress: string,
+  adminBytecode: string,
+) {
+  const { deployData, simulateDeploy } = await getSimulatedData(
+    hre,
+    ProxyAdminFactory,
+    opts,
+    adminAddress,
+    adminBytecode,
+  );
+  const manifestAdminAddress = await fetchOrDeployAdmin(deployData.provider, simulateDeploy, opts);
+  if (adminAddress !== manifestAdminAddress) {
+    logWarning(
+      `Imported proxy with admin at '${adminAddress}' which differs from previously deployed admin '${manifestAdminAddress}'`,
+      [
+        `The imported proxy admin is different from the proxy admin that was previously deployed on this network. This proxy will not be upgradable directly by the plugin.`,
+        `To upgrade this proxy, use the prepareUpgrade or defender.proposeUpgrade function and then upgrade it using the admin at '${adminAddress}'.`,
+      ],
+    );
+  }
+}
+
+export async function simulateDeployImpl(
+  hre: HardhatRuntimeEnvironment,
+  ImplFactory: ContractFactory,
+  opts: Options,
+  implAddress: string,
+  runtimeBytecode: string,
+) {
+  const { deployData, simulateDeploy } = await getSimulatedData(hre, ImplFactory, opts, implAddress, runtimeBytecode);
+  await fetchOrDeploy(deployData.version, deployData.provider, simulateDeploy, opts, true);
+}
+
+/**
+ * Gets data for a simulated deployment of the given contract to the given address.
+ */
+async function getSimulatedData(
+  hre: HardhatRuntimeEnvironment,
+  ImplFactory: ContractFactory,
+  opts: Options,
+  implAddress: string,
+  runtimeBytecode?: string,
+) {
+  const deployData = await getDeployData(hre, ImplFactory, opts);
+  const simulateDeploy = async () => {
+    const abi = ImplFactory.interface.format(FormatTypes.minimal) as string[];
+    const deployment = Object.assign({ abi });
```

Comment:
This looks a little weird, what's the goal? Why not just inline `abi` below in the object below?